index.ts

```diff
@@ -14,23 +15,35 @@ abstract class NgComponent<Props, State> {
     }
   */
   // nb: this method is explicity exposed for unit testing
-  public $onChanges(changes: object) {
+  $onChanges(changes: object) {
     const oldProps = mapValues<{}, Props>(changes, 'previousValue')
     const newProps = mapValues<{}, Props>(changes, 'currentValue')
+    if (!this.__isFirstRender && !this.didPropsChange(newProps, oldProps)) return
+    this.lifecycle(newProps, {} as State)
+  }
 
+  // note: this isn't meant to exactly replicate React's setState (e.g. http://stackoverflow.com/q/28922275/7816712)
+  protected setState(newState: State) {
+    if (this.__setStateTriggersLifecycle) {
+      setTimeout(() => this.lifecycle({} as Props, newState), 0)
```

Comment:
set comment above - it's sometimes sync or async... haven't looked into exactly how this behavior works